consensus_test.go

```diff
@@ -444,6 +525,294 @@ func TestTransition_RoundChangeState_MaxRound(t *testing.T) {
+	m.state.view = ViewMsg(1, 2)
+
+	// Send message from the past (it should be discarded)
+	m.emitMsg(&MessageReq{
+		From: "A",
+		Type: MessageReq_Prepare,
+		View: ViewMsg(1, 1),
+	})
+	// Send future message
+	m.emitMsg(&MessageReq{
+		From: "B",
+		Type: MessageReq_Prepare,
+		View: ViewMsg(2, 3),
+	})
+
+	m.runCycle(context.Background())
+	m.expect(expectResult{
+		state:       RoundChangeState,
+		round:       2,
+		sequence:    1,
+		prepareMsgs: 0,
+		commitMsgs:  0,
+		outgoing:    0})
+}
+
+func TestTransition_CommitState_DoneState(t *testing.T) {
+	m := newMockPbft(t, []string{"A", "B", "C"}, "A")
+	m.state.view = ViewMsg(1, 0)
+	m.state.proposer = "A"
+	m.setState(CommitState)
+
+	m.runCycle(context.Background())
+
+	m.expect(expectResult{
+		sequence: 1,
+		state:    DoneState,
+	})
+}
+
+func TestTransition_CommitState_RoundChange(t *testing.T) {
+	m := newMockPbft(t, []string{"A", "B", "C"}, "A")
+	m.state.view = ViewMsg(1, 0)
+	m.setState(CommitState)
+
+	m.runCycle(context.Background())
+
+	m.expect(expectResult{
+		sequence: 1,
+		state:    RoundChangeState,
+		err:      errFailedToInsertProposal,
+	})
+	assert.True(t, m.IsState(RoundChangeState))
+}
+
+func TestExponentialTimeout(t *testing.T) {
+	testCases := []struct {
+		description string
+		round       uint64
+		expected    time.Duration
+	}{
+		{"for round 0", 0, defaultTimeout + (1 * time.Second)},
+		{"for round 1", 1, defaultTimeout + (2 * time.Second)},
+		{"for round 2", 2, defaultTimeout + (4 * time.Second)},
+		{"for round 8", 8, defaultTimeout + (256 * time.Second)},
+		{"for round 9", 9, maxTimeout},
+		{"for round 10", 10, maxTimeout},
+		{"for round 34", 34, maxTimeout},
+	}
+
+	for _, tc := range testCases {
+		tc := tc // rebind tc into this lexical scope
+		t.Run(tc.description, func(t *testing.T) {
+			t.Parallel()
+			p := Pbft{state: &currentState{view: &View{Round: tc.round}}}
+			timeout := p.exponentialTimeout()
+			require.Equal(t, tc.expected, timeout, fmt.Sprintf("timeout should be %s", tc.expected))
+		})
+	}
+}
+
+func TestDoneState_RunCycle_Panics(t *testing.T) {
+	m := newMockPbft(t, []string{"A", "B", "C"}, "A")
+	m.state.view = ViewMsg(1, 0)
+	m.SetState(DoneState)
+
+	assert.Panics(t, func() { m.runCycle(context.Background()) })
+}
+
+func TestPbft_Run(t *testing.T) {
+	m := newMockPbft(t, []string{"A", "B", "C"}, "A")
+	m.state.view = ViewMsg(1, 0)
+	m.setProposal(&Proposal{
+		Data: mockProposal,
+		Time: time.Now(),
+	})
+
+	// Prepare messages
+	m.emitMsg(&MessageReq{
+		From: "A",
+		Type: MessageReq_Prepare,
+		View: ViewMsg(1, 0),
+	})
+	m.emitMsg(&MessageReq{
+		From: "B",
+		Type: MessageReq_Prepare,
+		View: ViewMsg(1, 0),
+	})
+	m.emitMsg(&MessageReq{
+		From: "C",
+		Type: MessageReq_Prepare,
+		View: ViewMsg(1, 0),
+	})
+
+	// Jump out from a state machine loop straight away
+	ch := make(chan struct{})
+	go func() {
+		close(ch)
+		for {
+			if m.getState() == AcceptState {
+				m.cancelFn()
+				return
+			}
+		}
+	}()
+	<-ch
```

Comment:
What is this channel doing here?

Reply:
Purpose of this channel is to make sure that go routine started execution, before state machine began to run, in order to be able to cancel it immediately (while we are still in `AcceptState`). Any ideas on how to accomplish this behavior, but more elegant are welcome. :slightly_smiling_face:

Reply:
This is resolved right? I cannot find it anymore.

Reply:
Channel remained in the code, I have just renamed it to `waitSignal`. The rationale I did it that way is described in previous comment (https://github.com/0xPolygon/pbft-consensus/pull/15#discussion_r787940759). I have tried to remove channel, but test becomes unstable and it depends when go routine begins executing.
Also the reason I haven't changed it to `runCycle`, but left `Run` function instead is because I wanted to cover main state machine loop as well in at least one test.

We can change test in some other PR if you want, but I am running out of ideas what to do here. 🤔